RNMT-3066 Automatic Tests

ines-sousa opened the pull request:
## Description
<!--- Describe your changes in detail -->
Execution of two P0's (and one P2) for the Contacts Plugin Sample App. The tests include a setup and the test itself. It should also include a teardown (the contact created should be deleted after the test), however, for now, this won't be possible.
Tests were executed with Android 9 emulator.
Test Plan: https://docs.google.com/document/d/1MAvo-wzbbkz9VVLc5yoaQlershXUEm6LREoJSiLMTes/edit?pli=1#heading=h.bplj4oklubeb

## Context
<!--- Place the link to the issue here preceded by either 'Closes' OR 'Fixes' -->
https://outsystemsrd.atlassian.net/browse/RNMT-3066 
<!--- Why is this change required? What problem does it solve? -->

## Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply -->
- [ ] Feature (change which adds functionality)
    - [ ] BREAKING CHANGE (existing functionality will not work as expected due to new feature)
- [ ] Fix (change which fixes an issue)
    - [ ] BREAKING CHANGE (existing functionality will not work as expected due to fix)
- [ ] Performance (change which improves performance)
    - [ ] BREAKING CHANGE (existing functionality will not work as expected due to performance improvement)
- [ ] Refactor (non-breaking change that is neither feature, fix nor performance)
- [ ] Style (non-breaking change that only affects formatting and/or white-space)

## Components affected
- [ ] Android
- [ ] iOS
- [ ] JS
- [x] Appium

## Tests
<!--- Describe how you tested your changes in detail -->
<!--- Include details of your test environment if relevant -->

## Screenshots (if appropriate)
<!--- E.g. before change & after change -->

## Checklist
<!--- Go over all the following items and put an `x` in all the boxes that apply -->
- [x] Pull request title follows the format `RNMT-XXXX <title>`
- [ ] Tests have been created
- [ ] Code follows code style of this project
- [ ] CHANGELOG.md file is correctly updated
- [ ] Changes require an update to the documentation
	- [ ] Documentation has been updated accordingly

ines-sousa committed:
RNMT-3068 - Clone and Setup

Clone and setup of automation framework

ines-sousa committed:
test: RNMT-3067

Implement 2 P0's

sandrarodrigues76 committed:
tests:RNMT-3067

ines-sousa committed:
test: new apk, ipa and chromedriver 2.44

ines-sousa committed:
test: changes to import of Constants

The import of the constants (default_timeout and default_timeout_interval) has been changed

ines-sousa committed:
test: add screen objects

ines-sousa committed:
test: Add another test to contact.spec.ts

ines-sousa committed:
test: add more chromedrivers to the folder

sandrarodrigues76 committed:
test: import constant values (missing files: AndroidUtils and iOSUtils)

sandrarodrigues76 committed:
tests: small fixes

ines-sousa committed:
test: change priority in test

ines-sousa committed:
Revert "Merge branch 'RNMT-3066-(Automatic-Tests)' of https://github.com/OutSystems/cordova-plugin-contacts into RNMT-3066-(Automatic-Tests)"

This reverts commit bf39cbc8eee54ee4061e74a965946235cc7b8d0d, reversing
changes made to 77e5dd06c3b75b5f288ddebf0d941375568cd0eb.

ines-sousa committed:
test: update ContactScreen.ts with all page object

ines-sousa committed:
chore: updates .gitignore file

ines-sousa committed:
test: fixes missing constants

ines-sousa committed:
chore: updates .gitignore file

ines-sousa committed:
chore: updates .gitignore files

ines-sousa committed:
chore: updates .gitignore

ines-sousa committed:
test: updates ContactsScreen.js page objects

ines-sousa committed:
test: adds the second P0's test scenario to contact.spec.ts

ines-sousa committed:
chore: update .gitignore to ignore .tar files

ines-sousa committed:
test: adds two components to the page object ContactsScreen.ts

ines-sousa committed:
test: adds before all/ after all instead of before each/after each

And makes some changes: some parts of the code went to the before all instead of being inside the test (in the before all section, a contact that is needed for every test is created)

EiyuuZack committed:
chore: remove generated files

They are not supposed to be on the repo.

References https://outsystemsrd.atlassian.net/browse/RNMT-3066

EiyuuZack committed:
build: update package.json and package-lock.json

References https://outsystemsrd.atlassian.net/browse/RNMT-3066

ines-sousa committed:
build: update .gitignore

Remove generated files that should not be on the repository.

References https://outsystemsrd.atlassian.net/browse/RNMT-3066

ines-sousa committed:
test: fix typo

References https://outsystemsrd.atlassian.net/browse/RNMT-3066

ines-sousa committed:
test: update chromedriver to version 75

Requires at least Chrome version 75.0.3770.140 to be installed in
emulator.

References https://outsystemsrd.atlassian.net/browse/RNMT-3066

EiyuuZack commented on os-appium-tests/README.md:
```diff
@@ -0,0 +1,37 @@
+# OutSystems Appium Tests Framework
```
This documentation does not mention those steps you showed me about setting up chromedriver. Could you add instructions, even if it is for Windows?

ines-sousa replied:
Done!

EiyuuZack commented on os-appium-tests/config/wdio.android.app.conf.js:
```diff
@@ -0,0 +1,25 @@
+        // The defaults you need to have in your config
+        automationName: 'UiAutomator2',
+        platformName: 'Android',
+        deviceName: 'Nexus',
```
Is this name relevant for anything?

ines-sousa replied:
No :P, I changed to: deviceName: "Android"

pmaroco commented on os-appium-tests/tests/constants.ts:
```diff
@@ -0,0 +1,2 @@
+export const DEFAULT_TIMEOUT = 10000;
+export const DEFAULT_TIMEOUT_INTERVAL = 1000;
```
Can you explain the difference between DEFAULT_TIMEOUT and DEFAULT_TIMEOUT_INTERVAL? Can we have a better name?

pmaroco commented on os-appium-tests/tests/helpers/NativeAlert.ts:
```diff
@@ -0,0 +1,62 @@
+     */
+    public static waitForIsShown(isShown = true, driver): void {
+        const selector = driver.isAndroid ? SELECTORS.ANDROID.ALERT_DIALOG : SELECTORS.IOS.ALERT;
+        $(selector).waitForExist(DEFAULT_TIMEOUT, !isShown); // ISO: commented because native.alert wasn't working with it
```
where is the comment? 🤔

pmaroco commented on os-appium-tests/tests/constants.ts:
```diff
@@ -0,0 +1,2 @@
+export const DEFAULT_TIMEOUT = 10000;
+export const DEFAULT_TIMEOUT_INTERVAL = 1000;
```
Can you explain the difference between DEFAULT_TIMEOUT and DEFAULT_TIMEOUT_INTERVAL? Can we have a better name?

sandrarodrigues76 replied:
@pmaroco  These two values came from the framework - we only make it work adding the export ...  :(, it's no excuse but the comment imo should be changed or improved at the framework

pmaroco commented on os-appium-tests/tests/helpers/NativeAlert.ts:
```diff
@@ -0,0 +1,62 @@
+const SELECTORS = {
+    ANDROID: {
+        ALERT_BUTTON: '*//android.widget.Button[@text="{BUTTON_TEXT}"]',
+        ALERT_DIALOG: '*//android.widget.LinearLayout[@resource-id="com.android.packageinstaller:id/dialog_container"]', // Sandra
```
Can you clean the leftover comments from the project? `// Sandra` is not quite informative :)

pmaroco commented on os-appium-tests/tests/helpers/NativeAlert.ts:
```diff
@@ -0,0 +1,62 @@
+        ALERT_BUTTON: '*//android.widget.Button[@text="{BUTTON_TEXT}"]',
+        ALERT_DIALOG: '*//android.widget.LinearLayout[@resource-id="com.android.packageinstaller:id/dialog_container"]', // Sandra
+        ALERT_TITLE: '*//android.widget.TextView[@resource-id="com.android.packageinstaller:id/permission_message"]', // ANDROID 9
+        // ALERT_TITLE: '*//android.widget.TextView[@resource-id="android:id/alertTitle"]', //BEFORE ANDROID 9
```
So we should use selector `ALERT_TITLE: '*//android.widget.TextView[@resource-id="android:id/alertTitle"` is we are testing versions before Android 9? If so, can we know the Android version using to tests on runtime and change the selector content dynamically? 

This would allow us to quickly change the .ipa without having to refactor the code :)

sandrarodrigues76 replied:
Great suggestion, that refactor should be done at the framework file @ines-sousa

pmaroco commented on os-appium-tests/tests/helpers/NativeAlert.ts:
```diff
@@ -0,0 +1,62 @@
+     * ANDROID:
+     *  Use the text of the button, provide a string and it will automatically transform it to uppercase
+     *  and click on the button
+     *
```
Great explanation! 🙆

EiyuuZack replied:
🔝

pmaroco reviewed (changes requested):
Good job overall! 👍

sandrarodrigues76 commented on os-appium-tests/README.md:
```diff
@@ -0,0 +1,37 @@
+    * either by using the command line or launching the server with the appium desktop application
+    * have a device connected or emulator available
+2. Run the command `npm run android`
+
```
or Run the command 'npm run iOS"

sandrarodrigues76 commented on os-appium-tests/README.md:
```diff
@@ -0,0 +1,37 @@
+1. create a run in aws
+1. upload your application (either _.ipa_ or _.apk_). Hit Next.
+1. upload the test bundle
+    * run the command `npm run bundleAws` and upload the generated `awsTests.zip` file.  
```
This is not working as expected we should remove by now and add the other set of steps

sandrarodrigues76 commented on os-appium-tests/awsConfiguration.yml:
```diff
@@ -0,0 +1,84 @@
+version: 0.1
```
even this file is prepared to android , iOS the configurations are different we should make it explicit

sandrarodrigues76 reviewed (commented):
We should move the directories apps and "chromedriver" to the level above, the bundle uses everything that exists at this level

sandrarodrigues76 commented on os-appium-tests/config/wdio.android.app.conf.js:
```diff
@@ -0,0 +1,25 @@
+const { join } = require('path'); // added by Ines & Sandra
```
remove the comment pls

sandrarodrigues76 commented on os-appium-tests/tests/screenobjects/SampleScreen.ts:
```diff
@@ -0,0 +1,23 @@
+import * as AndroidUtils from '../helpers/AndroidUtils';
```
we must delete this file.

sandrarodrigues76 commented on os-appium-tests/tests/specs/contact.spec.ts:
```diff
@@ -0,0 +1,102 @@
+/*
+After each test scenario (or each suite of tests), the test should have a teardown where all contacts created are deleted
+However, this one doesn't have teardown, because the plugin can't delete contacts and, with appium, we can't access the native contacts and delete them
+*/
```
"This can impact other test files, we should be precise with expects" - I think we should add some explanation about the impact of leaving the contacts there.

sandrarodrigues76 commented on os-appium-tests/tests/specs/contact.spec.ts:
```diff
@@ -0,0 +1,102 @@
+        // The expected result is for the contact to be created (message text = true)
+        expect(ContactsScreen.getFeedbackMessage().getText()).toEqual('Success: True');
+
+        // discover a way to validate the number in the contacts
```
we need this validation for this test case?

ines-sousa replied:
The test consists in creating a new contact, so I think the expected result is the message saying the contact was created. Of course this is not the best approach, we should verified the contact was created in the phone contact list. But, given the difficulty in accessing the contacts list on the phone, I thought this could be a solution.

sandrarodrigues76 commented on os-appium-tests/tests/specs/contact.spec.ts:
```diff
@@ -0,0 +1,102 @@
+        // Test: click to create the contact
+        ContactsScreen.getAddContactButton().click();
+
+        // The expected result is for the contact to be created (message text = true)
```
hummm... are we validating what the description is mention? "with the same number", can't we validate that?

ines-sousa replied:
Yes!! You're right, in the next test case (Find Contact by First Name) I will add a validation to see if the new phone number was added to the contact. What do you think?
PS: the name of the test case was wrong, it should be: "Add contact with different number"
![image](https://user-images.githubusercontent.com/47888982/62611787-2a285100-b8fe-11e9-8bec-1b3ad10c2c1a.png)

usernuno committed:
test: fix appium tests device farm config

luissilvaos committed:
test: refactor screen object ContactScreen

luissilvaos committed:
test: fix native alert helper

luissilvaos committed:
test: remove debugger on afterTest event

luissilvaos committed:
test: add PermisionAlert helper for Android Permissions

luissilvaos committed:
test: fix tests for the contacts sample app

luissilvaos committed:
test: add permission alert helper for iOS

luissilvaos committed:
test: fix contacts spec

luissilvaos committed:
test: change platform configurations

luissilvaos committed:
test: remove sample screen object

luissilvaos committed:
test: remove unused helper from the contacts spec

luissilvaos committed:
test: update application packages and chromedriver files

luissilvaos committed:
test: update gitignore to allow the zip file with the iOS.app file for local tests

luissilvaos committed:
test: add missing documentation on PermissionAlert.ts

luissilvaos committed:
test: change AWS Configuration files for Android and iOS

Chuckytuh commented on os-appium-tests/config/wdio.android.app.conf.js:
```diff
@@ -0,0 +1,25 @@
+const { join } = require('path');
```
does AWS Device farm runner support ES6 keywords such as const? What about object destructuring (`{join} = require("path")`?

Chuckytuh commented on os-appium-tests/config/wdio.ios.app.conf.js:
```diff
@@ -0,0 +1,21 @@
+const { config } = require('./wdio.shared.conf');
+const { join } = require('path'); // added by Ines & Sandra
```
added by Ines & Sandra <3

Chuckytuh commented on os-appium-tests/config/wdio.shared.conf.js:
```diff
@@ -0,0 +1,76 @@
+    // Specs
+    // ============
+    specs: [
+        './tests/specs/**/*.spec.ts'
```
Why are specs defined here but also in https://github.com/OutSystems/cordova-plugin-contacts/blob/ce9666ba82ecd4bc5974b18da32dc9d61385e1da/os-appium-tests/config/wdio.devicefarm.conf.js ?

luissilvaos replied:
This is the default configuration that comes from the framework, and are used for local developments. For the AWS Device Farm, we are using the generated js files.

Chuckytuh replied:
Ah I see, thanks!

luissilvaos committed:
test: refactor contacts spec to look for success cards instead of feedback messages

luissilvaos committed:
test: remove unused code and add screen validator to last test

ines-sousa committed:
test: change readme file with new .yml files

pmaroco reviewed (approved):
I'm giving you the 💚 but I think there are still some comments in the project that should be removed

ines-sousa committed:
test: update comment in contact.spec.ts about the impact of not deleting tests

EiyuuZack reviewed (approved):
I'll approve, but please check my comments. Also are chromedriver 75 and 76 both required? Can we keep just one?

ines-sousa committed:
test: update contact.spec.ts

Improve typing (spaces, commas, etc)
Change second test case name (it was wrongly "same number")
Add validation in test "Find contact by first name"

https://outsystemsrd.atlassian.net/browse/RNMT-3066

ines-sousa committed:
test: update README.md

add steps to solve the chromedriver in Android problem (appium is not recognizing the chromedriver)

ines-sousa committed:
test: update wdio.android.app.config.js (device name is now "Android")

https://outsystemsrd.atlassian.net/browse/RNMT-3066
